packages/calypso-e2e/src/lib/pages/editor-page.ts

```diff
@@ -563,39 +564,51 @@ export class EditorPage {
 	//#region Publish, Draft & Schedule
 
 	/**
-	 * Publishes the post or page.
+	 * Publishes the post or page and returns the resulting URL.
+	 *
+	 * If the optional parameter `visit` parameter is specified, the page is navigated
+	 * to the published article.
 	 *
 	 * @param {boolean} visit Whether to then visit the page.
 	 * @returns {URL} Published article's URL.
 	 */
 	async publish( { visit = false }: { visit?: boolean } = {} ): Promise< URL > {
-		// Click on the main publish action button on the toolbar.
-		await this.editorToolbarComponent.clickPublish();
+		const publishButtonText = await this.editorToolbarComponent.getPublishButtonText();
+		const actionsArray = [];
+
+		// Intercept the response from the WordPress.com REST API.
+		actionsArray.push( this.page.waitForResponse( /sites.*\/(posts|pages)\/.*/ ) );
+
+		// Every publish action requires at least one click on the EditorToolbarComponent.
+		actionsArray.push( this.editorToolbarComponent.clickPublish() );
 
-		if ( await this.editorPublishPanelComponent.panelIsOpen() ) {
-			// Invoke the second stage of the publish step which handles the
-			// publish checklist panel if it is present.
-			await this.editorPublishPanelComponent.publish();
+		// Determine whether the post/page is yet to be published or the post/page
+		// is merely being updated.
+		// If not yet published, a second click on the EditorPublishPanelComponent
+		// is added to the array of actions.
+		if ( publishButtonText.toLowerCase() === 'publish' ) {
+			actionsArray.push( this.editorPublishPanelComponent.publish() );
```

Comment:
This change, while seemingly unrelated to the task of eliminating the use of Editor handles, is actually quite important.

In its current form, the `publish` method uses entirely UI-driven methods to publish then obtain the published post/page's URL.

Not only is this possibly flaky as the post/page publish toast fades after some time, the post-publish confirmation panel can also be auto-dismissed due to issues with some blocks (eg. https://github.com/Automattic/wp-calypso/issues/54421).

By intercepting the response to the publishing endpoint, the process to obtain the post/page's URL is made much more reliable.